src/Compilers/Core/Portable/Compilation/Compilation.cs

```diff
@@ -839,6 +839,20 @@ public INamedTypeSymbol CreateTupleTypeSymbol(ImmutableArray<ITypeSymbol> elemen
 
         protected abstract INamedTypeSymbol CommonCreateTupleTypeSymbol(ImmutableArray<ITypeSymbol> elementTypes, ImmutableArray<string> elementNames);
 
+        /// <summary>
+        /// Returns a new INamedTypeSymbol with the given underlying type and (optional) element names.
+        /// </summary>
+        /// <remarks>
+        /// Since VB doesn't support tuples yet, this call will fail in a VB compilation.
+        /// Also, the underlying type needs to be tuple-compatible.
+        /// </remarks>
+        public INamedTypeSymbol CreateTupleTypeSymbol(INamedTypeSymbol underlyingType, ImmutableArray<string> elementNames = default(ImmutableArray<string>))
```

Comment:
> That does not make sense to me. Why forbid creating a tuple from an underlying type with names? If I can create a tuple from underlying type, then I need to control its element names just like I would for any other tuple.

That's not the way the language works.

> You wrote that isn't something you can do in the language, but what if I create 'a.dll' with a ValueTuple type and declaring an API returning a tuple with names from that? Isn't that a tuple where I control the underlying type and also specify names?

No. In that case you use the second API and specify the underlying assembly.

> Incidentally, the refactoring that is prompting this API should work for any underlying type and carry element names through. How can I round-trip otherwise?

No refactoring should ever have a type that neither API could create.